File: Boba.Core/Boolean.fs
namespace Boba.Core

module Boolean =
    
    open System.Diagnostics
    open Common
    
    /// Represents a Boolean equation with variables and the standard Boolean constants. How these constants and variables are
    /// interpreted is up to the consumer of this data type. For Boba, though, we include dotted variables, which are generally
    /// substituted with a sequence that is converted to nested disjunctions. There is also a distinction between rigid and flexible
    /// variables: only flexible variables are considered 'free', and thus only they will be included in the unifier during unification.
    /// This enables us to do 'boolean matching' by setting one side of the equation to rigid, and the other flexible.
    [<DebuggerDisplay("{ToString()}")>]
    type Equation =
        | BTrue
        | BFalse
        | BVar of string
        | BDotVar of string
        | BRigid of string
        | BDotRigid of string
        | BNot of Equation
        | BAnd of Equation * Equation
        | BOr of Equation * Equation
        override this.ToString () =
            match this with
            | BTrue -> "1"
            | BFalse -> "0"
            | BVar n -> n
            | BDotVar n -> $"{n}..."
            | BRigid n -> $"{n}*"
            | BDotRigid n -> $"{n}*..."
            | BNot b ->
                match b with
                | BAnd _ -> $"!({b})"
                | BOr _ -> $"!({b})"
                | _ -> $"!{b}"
            | BAnd (l, r) ->
                match l, r with
                | BOr _, BOr _ -> $"({l})∧({r})"
                | BOr _, _ -> $"({l})∧{r}"
                | _, BOr _ -> $"{l}∧({r})"
                | _, _ -> $"{l}∧{r}"
            | BOr (l, r) ->
                match l, r with
                | BAnd _, BAnd _ -> $"({l})∨({r})"
                | BAnd _, _ -> $"({l})∨{r}"
                | _, BAnd _ -> $"{l}∨({r})"
                | _, _ -> $"{l}∨{r}"

    type MinTermRow = { Name: Set<int>; Row: List<int>; }

    [<Literal>]
    let QmcTrue = 1
    [<Literal>]
    let QmcFalse = 0
    [<Literal>]
    let QmcDash = 2

    let mkNot b =
        match b with
        | BFalse -> BTrue
        | BTrue -> BFalse
        | BNot s -> s
        | _ -> BNot b

    let mkNotV v = mkNot (BVar v)
    
    let mkAnd l r =
        match l, r with
        | BTrue, _ -> r
        | _, BTrue -> l
        | BFalse, _ -> BFalse
        | _, BFalse -> BFalse
        | _ when l = r -> l
        | _ when l = BNot r -> BFalse
        | _ when BNot l = r -> BFalse
        | _, BOr (sl, sr) when l = sl || l = sr -> l
        | BOr (sl, sr), _ when r = sl || r = sr -> r
        | _ -> BAnd (l, r)

    let mkAndV l r = mkAnd (BVar l) (BVar r)

    let mkOr l r =
        match l, r with
        | BTrue, _ -> BTrue
        | _, BTrue -> BTrue
        | BFalse, _ -> r
        | _, BFalse -> l
        | _ when l = r -> l
        | _ when l = BNot r -> BTrue
        | _ when BNot l = r -> BTrue
        | _, BAnd (sl, sr) when l = sl || l = sr -> l
        | BAnd (sl, sr), _ when r = sl || r = sr -> r
        | _ -> BOr (l, r)

    let mkOrV l r = mkOr (BVar l) (BVar r)

    let mkXor l r =
        mkOr (mkAnd l (mkNot r)) (mkAnd (mkNot l) r)

    /// Make all variables in the equation flexible.
    let rec flexify eqn =
        match eqn with
        | BRigid n -> BVar n
        | BDotRigid n -> BDotVar n
        | BNot b -> BNot (flexify b)
        | BAnd (l, r) -> BAnd (flexify l, flexify r)
        | BOr (l, r) -> BOr (flexify l, flexify r)
        | _ -> eqn

    /// Make all variables in the equation rigid.
    let rec rigidify eqn =
        match eqn with
        | BVar n -> BRigid n
        | BDotVar n -> BDotRigid n
        | BNot b -> BNot (rigidify b)
        | BAnd (l, r) -> BAnd (rigidify l, rigidify r)
        | BOr (l, r) -> BOr (rigidify l, rigidify r)
        | _ -> eqn
    
    /// Compute the set of free variables in the equation.
    let rec free eqn =
        match eqn with
        | BVar n -> Set.singleton n
        | BDotVar n -> Set.singleton n
        | BNot b -> free b
        | BAnd (l, r) -> Set.union (free l) (free r)
        | BOr (l, r) -> Set.union (free l) (free r)
        | _ -> Set.empty
    
    let rec freeWithCtor eqn =
        match eqn with
        | BVar n -> Map.add n BVar Map.empty
        | BDotVar n -> Map.add n BDotVar Map.empty
        | BNot b -> freeWithCtor b
        | BAnd (l, r) -> mapUnion fst (freeWithCtor l) (freeWithCtor r)
        | BOr (l, r) -> mapUnion fst (freeWithCtor l) (freeWithCtor r)
        | _ -> Map.empty

    /// Replace the given variable with sub in the target Boolean equation.
    let rec substitute var sub target =
        match target with
        | BVar n when n = var -> sub
        | BDotVar n when n = var -> sub
        | BNot b -> mkNot (substitute var sub b)
        | BAnd (l, r) -> mkAnd (substitute var sub l) (substitute var sub r)
        | BOr (l, r) -> mkOr (substitute var sub l) (substitute var sub r)
        | _ -> target

    /// Perform several successive substitution operations on the target equation.
    let applySubst subst target =
        Map.fold (fun eqn var sub -> substitute var sub eqn) target subst
    
    /// Combine two substitutions into a single substitution, such that applying them both separately has the same effect as applying the combined one.
    let mergeSubst subl subr = mapUnion fst subl subr

    /// Compute the truth set for the given equation as a list of BTrue or BFalse.
    let rec truthRow eqn freeVs =
        match freeVs with
        | [] ->
            if eqn = BTrue || eqn = BFalse then [eqn] else [eqn]
        | v :: vs ->
            let vTrue = truthRow (substitute v BTrue eqn) vs
            let vFalse = truthRow (substitute v BFalse eqn) vs
            List.append vTrue vFalse
    
    /// Compute the truth table for the given equation, returning a two dimensional list of the
    /// form [[T, T, ..., T],...,[F, F,..., T]] where the last element in each sublist is the
    /// truth value of the equation for that row.
    let rec truthTable eqn freeVs =
        match freeVs with
        | [] ->
            if eqn = BTrue || eqn = BFalse then [[eqn]] else [[eqn]]
        | v :: vs ->
            let vTrue = truthTable (substitute v BTrue eqn) vs |> List.map (fun r -> BTrue :: r)
            let vFalse = truthTable (substitute v BFalse eqn) vs |> List.map (fun r -> BFalse :: r)
            List.append vTrue vFalse

    /// Compute the truth set for the given equation as an integer where each bit is a truth value.
    /// For example, (a ∨ b) has truth row 1110 = 14. The integer representation allows us to handle
    /// up to 5 variables in an equation.
    let rec truthId eqn freeVs =
        match freeVs with
        | [] -> if eqn = BTrue then 1 else 0
        | v :: vs ->
            let vTrue = truthId (substitute v BTrue eqn) vs
            let vFalse = truthId (substitute v BFalse eqn) vs
            let shift = vs.Length + 1
            (vTrue <<< shift) ||| vFalse

    let eqnToQmc eqn =
        match eqn with
        | BTrue -> QmcTrue
        | BFalse -> QmcFalse
        | _ -> failwith $"Cannot convert term {eqn} to efficient QMC representation"
    
    let truthRowToInt =
        // take advantage of QMC representation to not have a branch for true/false here
        List.mapi (fun i e -> e <<< i) >> List.sum

    let rowNameEqual l r = l.Name = r.Name

    let minTermTrueCount minTerm =
        List.sumBy (fun e -> if e = QmcTrue then 1 else 0) minTerm.Row
    
    let rowDashCount =
        List.sumBy (fun e -> if e = QmcDash then 1 else 0)

    let tryGenerateComparedRow (left: List<int>) (right: List<int>) =
        let mutable difference = false
        let mutable tooMany = false
        let mutable row = []
        for i in 0..(left.Length-1) do
            if left.[i] <> right.[i]
            then
                // have we already seen a difference? if so, these rows are too different
                if difference
                then tooMany <- true
                else
                    difference <- true
                    row <- List.append row [QmcDash]
            else
                row <- List.append row [left.[i]]
        if difference && not tooMany
        then {| Compared = true; Row = row |}
        else {| Compared = false; Row = [] |}
    
    let compareRowAgainstOthers row otherRows expectedDashes =
        let zipped = [
            for o in otherRows do
                let cmp = tryGenerateComparedRow row.Row o.Row
                if cmp.Compared
                then yield ({ Name = Set.union row.Name o.Name; Row = cmp.Row }, o)
        ]
        List.unzip zipped

    let toTerm (free : List<string * (string -> Equation)>) ind eqn =
        let var, ctor = free.[ind]
        match eqn with
        | QmcTrue -> [ctor var]
        | QmcFalse -> [BNot (ctor var)]
        | _ -> []

    let toSum terms = List.fold mkAnd BTrue terms

    let toProduct sums = List.fold mkOr BFalse sums
    
    let primeImplicants minTerms =
        let rec implicantsIter steps primeImplicants remaining =
            let mutable newChecked = []
            let mutable newRemaining = []
            for i in 0 .. List.length remaining - 2 do
                for minTerm in remaining.[i] do
                    let implicants, matched = compareRowAgainstOthers minTerm remaining.[i+1] steps
                    if matched.Length > 0
                    then newChecked <- minTerm :: newChecked
                    newChecked <- List.append newChecked matched
                    newRemaining <- List.append implicants newRemaining
            newChecked <- List.distinctBy (fun c -> c.Row) newChecked
            newRemaining <- List.distinctBy (fun c -> c.Row) newRemaining
            let primes = [
                for g in remaining do
                    for e in g do
                        // if the row wasn't checked, it's a prime implicant
                        if not (List.exists (fun c -> e.Name = c.Name) newChecked)
                        then yield e]
            let grouped = List.groupBy minTermTrueCount newRemaining |> List.map snd
            checkContinue (steps + 1) (List.append primes primeImplicants) newChecked grouped
        and checkContinue steps primeImplicants checkedImplicants remaining =
            match checkedImplicants with
            | [] -> primeImplicants
            | _ -> implicantsIter steps primeImplicants remaining
        
        let grouped = List.groupBy minTermTrueCount minTerms |> List.map snd
        let implicants = implicantsIter 1 [] grouped
        // bigger names in front, so that de-duplicating chooses the one that covers the most minTerms
        let sorted = List.sortBy (fun imp -> -(Set.count imp.Name)) implicants
        List.distinctBy (fun c -> c.Row) sorted
    
    let essentialImplicants primes minTerms =
        let zipped = [
            for m in minTerms do
                let checks = List.filter (fun p -> Set.isSubset m.Name p.Name) primes
                if List.length checks = 1
                then (checks.[0], checks.[0].Name)
        ]
        let essentials, covered = List.unzip zipped
        let covered = Set.unionMany covered
        // determine which prime implicants and minterms are remaining to be investigated
        let remaining = List.filter (fun p -> not (List.exists (fun e -> p.Name = e.Name) essentials)) primes
        let uncovered = List.filter (fun m -> not (Set.isSubset m.Name covered)) minTerms
        essentials, covered, remaining, uncovered
    
    let productOfSums primes minTerms =
        [for m in minTerms ->
            [for p in primes do
                if Set.isSubset m.Name p.Name
                then yield p.Name]]
    
    let isSubset cmp test =
        List.forall (fun prime -> List.contains prime cmp) test
    
    let reduceByAbsorption sums =
        let mutable reduced = []
        let mutable sumCopy = List.map id sums
        while sumCopy.Length > 0 do
            let product = sumCopy.Head
            if not (List.exists (isSubset product) sumCopy || List.exists (isSubset product) reduced)
            then reduced <- product :: reduced
        reduced
    
    let productOfSumsToSumOfProducts product =
        let mutable sums = [List.head product]
        let mutable prodCopy = List.map id product
        while prodCopy.Length > 0 do
            let term = prodCopy.Head
            let mutable newSums = []
            for t in term do
                for s in sums do
                    if List.contains t s
                    then newSums <- s :: newSums
                    else newSums <- (t :: s) :: newSums
            sums <- newSums
        reduceByAbsorption sums

    let petricks primes minTerms =
        let product = productOfSums primes minTerms
        let sum = productOfSumsToSumOfProducts product
        let sorted = List.sortBy List.length sum
        [for name in sorted.Head -> List.filter (fun p -> p.Name = name) primes |> List.head]

    let minimize eqn =
        let fvs = freeWithCtor eqn |> Map.toList
        let truth = truthTable eqn (List.map fst fvs)

        // does it contain rigid terms?
        if not (List.forall (fun r -> r = BTrue || r = BFalse) (truthRow eqn (List.map fst fvs)))
        then eqn
        else

        // minTerms are elements of the truth table where the result is T.
        // It is convenient to not include the truth value element, so we remove
        // it from each row.
        let minTermRows = [for r in truth do if List.last r = BTrue then yield List.take (List.length r - 1) r]
        // convert to a more efficient representation, and one that supports dashes
        let qmcRows = [for r in minTermRows -> List.map eqnToQmc r]
        // give each minTerm a unique (but meaningful) name to efficiently refer to it later
        let namedMinTerms = [for r in qmcRows -> { Name = Set.singleton (truthRowToInt r); Row = r }]

        let primes = primeImplicants namedMinTerms

        let essentials, covered, remaining, uncovered = essentialImplicants primes namedMinTerms

        let finalImplicants =
            if uncovered <> []
            then
                let covering = petricks remaining uncovered
                List.append essentials covering
            else essentials
        
        // TODO: need to handle dotted vars here
        List.map (fun e -> List.mapi (toTerm fvs) e.Row |> List.concat |> toSum) finalImplicants |> toProduct

    /// Eliminate variables one by one by substituting them away
    /// and builds up a resulting substitution. Core of unification.
    let rec private successiveVariableElimination eqn vars =
        match vars with
        | [] ->
            // check whether any remaining rigid variables in the equation make it satisfiable
            if not (satisfiable eqn)
            then Some Map.empty
            else None
        | v :: vs ->
            let vFalse = substitute v BFalse eqn
            let vTrue = substitute v BTrue eqn
            let substRes = successiveVariableElimination (mkAnd vFalse vTrue) vs
            match substRes with
            | None -> None
            | Some subst ->
                let vsub = mkOr (applySubst subst vFalse) (mkAnd (BVar v) (mkNot (applySubst subst vTrue)))
                mergeSubst (Map.empty.Add(v, vsub)) subst |> Some

    /// Checks whether a given equation is satisfiable, i.e. whether there is a substitution of all variables to T or F that makes the equation T when evaluated.
    and satisfiable eqn =
        match minimize eqn with
        | BTrue -> true
        | BFalse -> false
        | _ ->
            let flexed = mkXor (flexify eqn) BTrue
            successiveVariableElimination flexed (free flexed |> Set.toList)
            |> Option.map (constant true)
            |> Option.defaultValue false

    /// Generate a most-general substitution that, when applied to both input equations,
    /// makes them equivalent boolean equations.
    let unify eqnl eqnr =
        // turn it into one equation to perform successive variable elimination
        let eqn = mkXor eqnl eqnr
        // find whichever side has fewer free variables, and eliminate those variables first
        // to yield a smaller unifier
        let lfree = free eqnl
        let rfree = free eqnr
        let sveVars =
            if rfree.Count < lfree.Count
            then List.append (List.ofSeq rfree) (List.ofSeq (Set.difference lfree rfree))
            else List.append (List.ofSeq lfree) (List.ofSeq (Set.difference rfree lfree))
        successiveVariableElimination eqn sveVars
        |> Option.map (mapValues minimize)

File: compiler/boolean/boolean.go
package boolean

import (
	"fmt"

	"github.com/glossopoeia/boba/compiler/util"
	"github.com/rjNemo/underscore"
	"golang.org/x/exp/constraints"
	"golang.org/x/exp/maps"
	"golang.org/x/exp/slices"
)

// Represents a Boolean equation with variables and the standard Boolean constants.
// How these constants and variables are interpreted is up to the consumer of this
// data type. For Boba, though, we include dotted variables, which are generally
// substituted with a sequence that is converted to nested disjunctions. There is
// also a distinction between rigid and flexible variables: only flexible variables
// are considered 'free', and thus only they will be included in the unifier during
// unification. This enables us to do 'boolean matching' by setting one side of the
// equation to rigid, and the other flexible.
type Equation[T constraints.Ordered] interface {
	fmt.Stringer
	// Helper method to construct the free set of variables efficiently
	// for each equation variant.
	freeAcc(map[T]Occurence, bool)
	// Make a subset of named rigids in the equation into flex variables.
	Flexify([]T) Equation[T]
	// Make a subset of flex variables into rigid variables in the equation.
	Rigidify([]T) Equation[T]
	// Substitute all occurences of the variables present in the substitution
	// with their assigned equation in the target Boolean equation.
	Substitute(Substitution[T]) Equation[T]
}

// Represents how many times a variable occurs in an equation when computing the
// set of free variables, and whether the variable occurred dotted or not.
type Occurence struct {
	Count  int
	Dotted bool
}

// A substitution on boolean equations, where T is the type of the variable names
// in the equation. Each T is mapped to a boolean sub-equation that will replace
// the variable in the equation and be simplified when subsitution is complete.
type Substitution[T constraints.Ordered] map[T]Equation[T]

type BTrue[T constraints.Ordered] struct{}

func (b BTrue[T]) String() string {
	return "T"
}

func (b BTrue[T]) freeAcc(acc map[T]Occurence, rigid bool) {}

func (b BTrue[T]) Flexify(subset []T) Equation[T] { return b }

func (b BTrue[T]) Rigidify(subset []T) Equation[T] { return b }

func (b BTrue[T]) Substitute(subst Substitution[T]) Equation[T] { return b }

type BFalse[T constraints.Ordered] struct{}

func (b BFalse[T]) String() string {
	return "F"
}

func (b BFalse[T]) freeAcc(acc map[T]Occurence, rigid bool) {}

func (b BFalse[T]) Flexify(subset []T) Equation[T] { return b }

func (b BFalse[T]) Rigidify(subset []T) Equation[T] { return b }

func (b BFalse[T]) Substitute(subst Substitution[T]) Equation[T] { return b }

type BVar[T constraints.Ordered] struct {
	Name   T
	Dotted bool
	Rigid  bool
}

func NewFlex[T constraints.Ordered](v T, dotted bool) Equation[T] {
	return BVar[T]{v, dotted, false}
}

func NewRigid[T constraints.Ordered](v T, dotted bool) Equation[T] {
	return BVar[T]{v, dotted, true}
}

func (v BVar[T]) String() string {
	if v.Dotted {
		if v.Rigid {
			return fmt.Sprintf("%v*...", v.Name)
		} else {
			return fmt.Sprintf("%v...", v.Name)
		}
	} else {
		if v.Rigid {
			return fmt.Sprintf("%v*", v.Name)
		} else {
			return fmt.Sprintf("%v", v.Name)
		}
	}
}

func (v BVar[T]) freeAcc(acc map[T]Occurence, rigid bool) {
	if v.Rigid == rigid {
		if occ, ok := acc[v.Name]; ok {
			acc[v.Name] = Occurence{occ.Count + 1, v.Dotted || occ.Dotted}
		} else {
			acc[v.Name] = Occurence{1, v.Dotted}
		}
	}
}

func (b BVar[T]) Flexify(subset []T) Equation[T] {
	if b.Rigid && slices.Contains(subset, b.Name) {
		return BVar[T]{b.Name, b.Dotted, false}
	}
	return b
}

func (b BVar[T]) Rigidify(subset []T) Equation[T] {
	if !b.Rigid && slices.Contains(subset, b.Name) {
		return BVar[T]{b.Name, b.Dotted, true}
	}
	return b
}

func (b BVar[T]) Substitute(subst Substitution[T]) Equation[T] {
	if !b.Rigid {
		if sub, ok := subst[b.Name]; ok {
			return sub
		}
	}
	return b
}

type BNot[T constraints.Ordered] struct {
	Elem Equation[T]
}

// Create a new Boolean Not equation from the equation to be negated,
// applying some minor simplifications if possible.
func NewNot[T constraints.Ordered](neg Equation[T]) Equation[T] {
	switch n := neg.(type) {
	case BTrue[T]:
		return BFalse[T]{}
	case BFalse[T]:
		return BTrue[T]{}
	case BNot[T]:
		return n.Elem
	default:
		return BNot[T]{neg}
	}
}

func (n BNot[T]) String() string {
	return fmt.Sprintf("!(%s)", n.Elem)
}

func (n BNot[T]) freeAcc(acc map[T]Occurence, rigid bool) {
	n.Elem.freeAcc(acc, rigid)
}

func (b BNot[T]) Flexify(subset []T) Equation[T] { return BNot[T]{b.Elem.Flexify(subset)} }

func (b BNot[T]) Rigidify(subset []T) Equation[T] { return BNot[T]{b.Elem.Rigidify(subset)} }

func (b BNot[T]) Substitute(subst Substitution[T]) Equation[T] {
	return NewNot(b.Elem.Substitute(subst))
}

type BAnd[T constraints.Ordered] struct {
	Left  Equation[T]
	Right Equation[T]
}

// Create a new Boolean And equation from two components, applying some minor simplifications
// if possible.
func NewAnd[T constraints.Ordered](left Equation[T], right Equation[T]) Equation[T] {
	switch left.(type) {
	case BTrue[T]:
		return right
	case BFalse[T]:
		return BFalse[T]{}
	default:
		switch right.(type) {
		case BTrue[T]:
			return left
		case BFalse[T]:
			return BFalse[T]{}
		default:
			return BAnd[T]{left, right}
		}
	}
}

func (a BAnd[T]) String() string {
	return fmt.Sprintf("(%s && %s)", a.Left, a.Right)
}

func (a BAnd[T]) freeAcc(acc map[T]Occurence, rigid bool) {
	a.Left.freeAcc(acc, rigid)
	a.Right.freeAcc(acc, rigid)
}

func (b BAnd[T]) Flexify(subset []T) Equation[T] {
	return BAnd[T]{b.Left.Flexify(subset), b.Right.Flexify(subset)}
}

func (b BAnd[T]) Rigidify(subset []T) Equation[T] {
	return BAnd[T]{b.Left.Rigidify(subset), b.Right.Rigidify(subset)}
}

func (b BAnd[T]) Substitute(subst Substitution[T]) Equation[T] {
	return NewAnd(b.Left.Substitute(subst), b.Right.Substitute(subst))
}

type BOr[T constraints.Ordered] struct {
	Left  Equation[T]
	Right Equation[T]
}

// Create a new Boolean Or equation from two components, applying some minor simplifications
// if possible.
func NewOr[T constraints.Ordered](left Equation[T], right Equation[T]) Equation[T] {
	switch left.(type) {
	case BTrue[T]:
		return BTrue[T]{}
	case BFalse[T]:
		return right
	default:
		switch right.(type) {
		case BTrue[T]:
			return BTrue[T]{}
		case BFalse[T]:
			return left
		default:
			return BOr[T]{left, right}
		}
	}
}

func (o BOr[T]) String() string {
	return fmt.Sprintf("(%s || %s)", o.Left, o.Right)
}

func (a BOr[T]) freeAcc(acc map[T]Occurence, rigid bool) {
	a.Left.freeAcc(acc, rigid)
	a.Right.freeAcc(acc, rigid)
}

func (b BOr[T]) Flexify(subset []T) Equation[T] {
	return BOr[T]{b.Left.Flexify(subset), b.Right.Flexify(subset)}
}

func (b BOr[T]) Rigidify(subset []T) Equation[T] {
	return BOr[T]{b.Left.Rigidify(subset), b.Right.Rigidify(subset)}
}

func (b BOr[T]) Substitute(subst Substitution[T]) Equation[T] {
	return NewOr(b.Left.Substitute(subst), b.Right.Substitute(subst))
}

// Create a new Boolean Xor equation from two components, applying some minor simplifications
// if possible.
func NewXor[T constraints.Ordered](l Equation[T], r Equation[T]) Equation[T] {
	return NewOr(NewAnd(l, NewNot(r)), NewAnd(NewNot(l), r))
}

// The number of occurrences of each distinct flexible variable in the equation.
func FreeFlex[T constraints.Ordered](eqn Equation[T]) map[T]Occurence {
	occ := make(map[T]Occurence)
	eqn.freeAcc(occ, false)
	return occ
}

// The number of occurrences of each distinct rigid variable in the equation.
func FreeRigid[T constraints.Ordered](eqn Equation[T]) map[T]Occurence {
	occ := make(map[T]Occurence)
	eqn.freeAcc(occ, true)
	return occ
}

// Eliminate variables one by one by substituting them away  and builds up a
// resulting substitution. If the equation is satisfiable, the result is nil.
// This is because unification in Boolean rings is matching when the equation
// equals false.
func successiveVariableElimination[T constraints.Ordered](eqn Equation[T], vars []T) Substitution[T] {
	if len(vars) == 0 {
		if !Satisfiable(Minimize(eqn)) {
			return Substitution[T]{}
		} else {
			return nil
		}
	}

	v := vars[0]
	vFalse := eqn.Substitute(Substitution[T]{v: BFalse[T]{}})
	vTrue := eqn.Substitute(Substitution[T]{v: BTrue[T]{}})
	subst := successiveVariableElimination(NewAnd(vFalse, vTrue), vars[1:])
	if subst != nil {
		vSub := NewOr(vFalse.Substitute(subst), NewAnd(NewFlex(v, false), NewNot(vTrue.Substitute(subst))))
		return util.MergeMaps(Substitution[T]{v: vSub}, subst, func(l, r Equation[T]) Equation[T] { return l })
	}
	return nil
}

// Checks whether a given equation is satisfiable, i.e. whether there is a substitution
// of all variables to T or F that makes the equation T when evaluated.
func Satisfiable[T constraints.Ordered](eqn Equation[T]) bool {
	switch eqn.(type) {
	case BTrue[T]:
		return true
	case BFalse[T]:
		return false
	default:
		flexed := NewXor(eqn.Flexify(maps.Keys(FreeRigid(eqn))), Equation[T](BTrue[T]{}))
		return successiveVariableElimination(flexed, maps.Keys(FreeFlex(flexed))) != nil
	}
}

// Generate a most-general substitution that, when applied to both input equations,
// makes them equivalent boolean equations.
func Unify[T constraints.Ordered](l Equation[T], r Equation[T]) Substitution[T] {
	// turn it into one equation to perform successive variable elimination
	eqn := NewXor(l, r)
	// find whichever side has fewer free variables, and eliminate those variables first
	// to yield a smaller unifier. Sort to maintain determinism of resulting unifier
	lfree := maps.Keys(FreeFlex(l))
	slices.Sort(lfree)
	rfree := maps.Keys(FreeFlex(r))
	slices.Sort(rfree)
	var sveVars []T
	if len(lfree) <= len(rfree) {
		sveVars = append(lfree, underscore.Difference(rfree, lfree)...)
	} else {
		sveVars = append(rfree, underscore.Difference(lfree, rfree)...)
	}

	subst := successiveVariableElimination(eqn, sveVars)
	for k, v := range subst {
		subst[k] = Minimize(v)
	}
	return subst
}
